File: src/main/java/com/developerphil/adbidea/compatibility/BackwardCompatibleGetter.java
package com.developerphil.adbidea.compatibility;

import org.joor.ReflectException;

/**
 * Abstracts the logic to call the current implementation and fall back on reflection for previous versions
 */
public abstract class BackwardCompatibleGetter<T> {

    public final T get() {
        try {
            return getCurrentImplementation();
        } catch (LinkageError error) {
            return getPreviousImplementation();
        } catch (Throwable e) {
            if (isReflectiveException(e)) {
                return getPreviousImplementation();
            } else {
                throw new RuntimeException(e);
            }
        }
    }

    private boolean isReflectiveException(Throwable t) {
        return t instanceof ClassNotFoundException ||
                t instanceof NoSuchFieldException ||
                t instanceof LinkageError ||
                t instanceof NoSuchMethodException ||
                t instanceof ReflectException
                ;
    }

    protected abstract T getCurrentImplementation() throws Throwable;

    protected abstract T getPreviousImplementation();
}

File: src/main/kotlin/com/developerphil/adbidea/compatibility/BackwardCompatibleGetter.kt
package com.developerphil.adbidea.compatibility

import org.joor.ReflectException

/**
 * Abstracts the logic to call the current implementation and fall back on reflection for previous versions
 */
abstract class BackwardCompatibleGetter<T> {
    fun get(): T {
        return try {
            getCurrentImplementation()
        } catch (error: LinkageError) {
            getPreviousImplementation()
        } catch (e: Throwable) {
            if (isReflectiveException(e)) {
                getPreviousImplementation()
            } else {
                throw RuntimeException(e)
            }
        }
    }

    private fun isReflectiveException(t: Throwable): Boolean {
        return t is ClassNotFoundException ||
                t is NoSuchFieldException ||
                t is LinkageError ||
                t is NoSuchMethodException ||
                t is ReflectException
    }

    abstract fun getCurrentImplementation() : T

    abstract fun getPreviousImplementation() : T
}